tokio/src/macros/join.rs

```diff
@@ -71,24 +71,50 @@ macro_rules! join {
         // the requirement of `Pin::new_unchecked` called below.
         let mut futures = ( $( maybe_done($e), )* );
 
+        // How many futures were passed to join!.
+        const FUTURE_COUNT: u32 = $crate::count!( $($count)* );
```

Comment:
If the `join!` macro previously worked with more than 64 branches, then this will break any such code. Please double check whether there the old version had a branch limit.

Reply:
We can pass 125 branches to the old version :thinking:

Reply:
I was trying to make the new `join!` version accept 125 branches too and kept hitting the macro recursion limit. I think it its because of the way the `join!` input is normalized.

```rust
    // ===== Normalize =====

    (@ { ( $($s:tt)* ) $($t:tt)* } $e:expr, $($r:tt)* ) => {
        $crate::join!(@{ ($($s)* _) $($t)* ($($s)*) $e, } $($r)*)
    };

    // ===== Entry point =====

    ( $($e:expr),* $(,)?) => {
        $crate::join!(@{ () } $($e,)*)
    };
```

I implemented `join!` as a proc macro, it was easier to avoid the recursion limit this way. Would a proc macro be ok?

Reply:
I would prefer to avoid a procedural macro if possible. In general, I recommend you ask about an implementation strategy first before implementing it, especially if it's a larger change.

Reply:
> I recommend you ask about an implementation strategy first before implementing it, especially if it's a larger change 

I think i read that in the CONTRIBUTING.md file, i will make sure to ask first next time. I chose to implement a proc macro without asking because i can use the practice if it is not accepted.

I went back to the version using just a declarative macro and was playing around with it the last couple of days but i'm pretty much stuck at the moment. 

I keep hitting the recursion limit when i pass 125 futures to `join`(the maximum supported without changing the recursion limit at the moment) because after `join!` normalizes its input, it is 126 levels deep in macro recursion and i need to call a macro `count!` once again.

Any ideas i could try?

Reply:
How about having the normalization count the number of branches?
```diff
         // One `_` for each branch in the `join!` macro. This is not used once
         // normalization is complete.
         ( $($count:tt)* )
+        // The expression `0+1+1+ ... +1` equal to the number of branches.
+        ( $($total:tt)* )
 
         // Normalized join! branches
         $( ( $($skip:tt)* ) $e:expr, )*
```
```diff
     // ===== Normalize =====
 
-    (@ { ( $($s:tt)* ) $($t:tt)* } $e:expr, $($r:tt)* ) => {
-        $crate::join!(@{ ($($s)* _) $($t)* ($($s)*) $e, } $($r)*)
+    (@ { ( $($s:tt)* ) ( $($n:tt)* ) $($t:tt)* } $e:expr, $($r:tt)* ) => {
+        $crate::join!(@{ ($($s)* _) ($($n)* + 1) $($t)* ($($s)*) $e, } $($r)*)
     };
 
     // ===== Entry point =====
 
     ( $($e:expr),* $(,)?) => {
-        $crate::join!(@{ () } $($e,)*)
+        $crate::join!(@{ () (0) } $($e,)*)
     };
 }
```
Then you could use the count in this way:
```Rust
// Safety: nothing must be moved out of `futures`. This is to satisfy
// the requirement of `Pin::new_unchecked` called below.
let mut futures = ( $( maybe_done($e), )* );

let mut skip_next_time: u32 = 0;

poll_fn(move |cx| {
    const COUNT: u32 = $($total)*;

    let mut is_pending = false;

    let mut to_run = COUNT;
    let mut skip = skip_next_time;
    skip_next_time = if skip+1 == COUNT { 0 } else { skip+1 };

    loop {
    $(
        if skip == 0 {
            if to_run == 0 {
                break;
            }
            to_run -= 1;

            // Extract the future for this branch from the tuple.
            let ( $($skip,)* fut, .. ) = &mut futures;

            // Safety: future is stored on the stack above
            // and never moved.
            let mut fut = unsafe { Pin::new_unchecked(fut) };

            // Try polling
            if fut.poll(cx).is_pending() {
                is_pending = true;
            }
        } else {
            skip -= 1;
        }
    )*
    }

    ...
}).await
```
Also, I think the same issue exists for the `try_join!` macro, so if the above works, it would be good to fix that one too.

Reply:
It works :+1: 

I should have thought about that :facepalm: